### ISSUE TITLE
CC-01-010 Evidence tampering via video upload (Critical)

### ISSUE THREAD

**Issue by cure53** [igarape/copcast-server#103]
It was found that the file upload feature fails to validate who the user is and therefore allows users to add videos under the account of any other user. This includes uploading the video itself as well as manipulation of the video timestamp.

These issues can be summarised as follows:

**Issue 1:** Video upload under the account of another user

Logged in as User ID 15 (abedroid1), a video was uploaded under User ID 23 (abedroid4) simply changing the username in the URL of the file upload:
        
**URL:**
https://pentest-api.mycopcast.com/videos/abedroid4

**Request:**
```
POST /videos/abedroid4 HTTP/1.1
Connection: close
ENCTYPE: multipart/form-data
Content-Type: multipart/form-data; boundary=---------------------------1469539718875
Authorization: Bearer eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJzdWIiOiI2ZWRlZWJhMi1lNTEwLTQ4MjQtODZmMi02NDEzMGY2YjM1MDgiLCJleHAiOjIwNjk1MjE1NTQxNTd9.pOsB1bB-WsG4iHFuh0oXzO3aA4pxgOBTZv4dsIilUiM
User-Agent: Dalvik/2.1.0 (Linux; U; Android 6.0.1; Nexus 5X Build/MMB29Q)
Host: pentest-api.mycopcast.com
Accept-Encoding: gzip
Content-Length: 49339


-----------------------------1469539718875
Content-Disposition: form-data; name="video"; filename="MjAxNi0wNy0yNVQxNjoyMjo1MC4xMjQrMDIwMA.mp4"
Content-Type: video/mpeg

[....]
```

The same video had been first uploaded under the abedroid1 account, and was successfully uploaded to the abedroid4 account (while being logged in as abedroid1) a few minutes later, as can be seen in the timestamps and file sizes of the videos on the server:

**Commands:**
```
cd /copcast/projects/copcast-server/videos/
ls -l 23/
```

**Output:**
```
total 356
-rw-rw-r-- 1 copcast copcast 344384 Jul 26 15:34 e608d09b-837d-47a8-99cc-3ffddc1d9698.m4a.enc
-rw-rw-r-- 1 copcast copcast  16176 Jul 26 15:34 e608d09b-837d-47a8-99cc-3ffddc1d9698.mp4.enc
```

The filesize confirms files are identical and the timestamp corresponds to the sequence of events, confirming the attack was successful:

**Command:**
`ls -l 15/`

**Output:**
```
total 356
-rw-rw-r-- 1 copcast copcast 344384 Jul 26 15:28 a5775ef7-9a73-438e-8b29-701fae6e987a.m4a.enc
-rw-rw-r-- 1 copcast copcast  16176 Jul 26 15:28 a5775ef7-9a73-438e-8b29-701fae6e987a.mp4.enc
```

**Issue 2:** Possible timestamp manipulation via filename

Snippet from the video upload above:
```
Content-Disposition: form-data; name="video"; 
filename="MjAxNi0wNy0yNVQxNjoyMjo1MC4xMjQrMDIwMA.mp4"
```

This decodes to the following timestamp:
`2016-07-25T16:22:50.124+020MA`

Which corresponds to the following code snippet, that retrieves the timestamp from the user-provided filename:

**File:**
lib/videos/index.js

**Affected Code:**
```
app.post('/videos/:user', auth.ensureUser, upload.single('video'), function (req, res) {

    var dateRecorded = null;
    var videoFile = req.file;
    if (videoFile.originalname.indexOf('-')>0) { // for handling older versions of copcast
      if (videoFile.originalname.indexOf(':') > 0) {
        dateRecorded = moment(videoFile.originalname.replace('.mp4', ''), 'YYYY-MM-DDTHH:mm:ss.SSSZ');
      } else {
        dateRecorded = moment(videoFile.originalname.replace('.mp4', ''), 'YYYY-MM-DD_HH-mm-ss');
      }
    } else {
      dateRecorded = moment(new Buffer(videoFile.originalname.replace('.mp4', ''), 'base64').toString('ascii'));
    }
```

It is recommended to use server-generated timestamps instead of client-supplied timestamps. In addition to this, the logged in user identity should be verified on each request to ensure a given user cannot read or upload data to another account.

**Comment by brunosiqueira** [comment]
@martelli leave a comment in slack so we can discuss more

**Comment by kamalg19** [comment]
@brunosiqueira, For app side, to protect timestamp manipulation via filename, we can always check if app is using network provided values for date/time settings. Let me know if you want me to handle this.

**Comment by brunosiqueira** [comment]
- [ ] create video structure (sql lite) in the client (register video name, start date, userlogin, hash)
- [ ] Register new record when video is created
- [ ] update record with hash when video is done
- [ ] when uploading the videos, send the table information to the server, the sessionId and a hash with { userId, sessionId, startTime, duration, sha256sum }
- [ ] after uploading, check the hash to verify if the video was tampered. 
- [ ] if tampered, send an email to the admin with the information about the video.

**Comment by brunosiqueira** [comment]
@kamalg19 can you help us with the "update record with hash when video is done" step?
We need to create the sha1 of the video + user login and register it in the database. I already created the video register when the recording starts. We need now to update this register with the hash to send to the server afterwards. The branch is feature_#103

**Comment by kamalg19** [comment]
@brunosiqueira While working on task "update record with hash when video is done", noticed few things:
1. Step "Register new record when video is created" was not working due to bug in WebRecorder.Builder. Fixed it.
2. Why we need two steps - first register video record in db and then update its hash value in db? Why not only one step to add record in db when video is done? 
Point 2 is important as our current code doesn't support updates of records(only write/get is supported). So, I need to make changes which will affect every place where we get data from db.

**Comment by brunosiqueira** [comment]
Makes sense, @kamalg19. You can register in the sqllite only once when the video is ready with all the information already. You can proceed.

**Comment by kamalg19** [comment]
@brunosiqueira Created pull request. Branch is feature_#103

Fixed
- Bug in WebRecorder.Builder
- Register video record to sqllitedb when video is done.
- Record contains date, file path, user login name, sha1 hash of video file + user name
